### PR TITLE
[MB-3718] Generate reference ID for payment service items

### PR DESCRIPTION
**Pull request by reggieriser**
## Description

This PR adds an automatically-generated reference ID for new payment service items (PSI).  In addition, this reference ID is used in the `N9` segment of the EDI, and is also returned as part of the JSON payload for payment service items.

As detailed in the [JIRA story](https://dp3.atlassian.net/browse/MB-3718), the format of the reference ID is: `<MTO Reference ID>-<enough of PSI ID to make it unique>`.  A minimum of 8 characters of the PSI ID are used.  More are added if necessary to make it unique, up to a maximum of 30 characters for the overall reference ID (a limit mandated by the max allowed size in the EDI).

## Reviewer Notes

- The story talks about creating a reference ID upon the approval of the payment service item, but after [some discussion](https://ustcdp3.slack.com/archives/CP6F568DC/p1601933781176800), we decided to just do it at creation.  This has the advantage of allowing us to make the field non-nullable.
- I use Pop's `BeforeCreate` callback to create a UUID (needed to compose the reference ID) and generate the reference ID itself.
- Note that the migration gets rid of all the existing payment requests (and related records) in staging/experimental (there should not be any in production yet).  We do this to avoid writing a complex SQL migration to "fix" any payment service items in deployed environments.

## Setup

1. `make db_dev_e2e_populate`
1. Use your favorite DB client and look at the `payment_service_items` table.  You should see that the new `reference_id` column is populated with an appropriate reference ID based on the format above.
1. `make server_run`
1. Save the following payload in `create_payment_request.json`:
    ```
    {
      "body": {
        "isFinal": false,
        "moveTaskOrderID": "9c7b255c-2981-4bf8-839f-61c7458e2b4d",
        "serviceItems": [
          {
            "id": "722a6f4e-b438-4655-88c7-51609056550d"
          },
          {
            "id": "ca9aeb58-e5a9-44b0-abe8-81d233dbdebf"
          },
          {
            "id": "6431e3e2-4ee4-41b5-b226-393f9133eb6c"
          },
          {
            "id": "fd6741a5-a92c-44d5-8303-1d7f5e60afbf"
          }
        ]
      }
    }
    ```
1. Create the payment request using the file above:
    ```
    go run ./cmd/prime-api-client --insecure create-payment-request --filename create_payment_request.json | jq
    ```
    The returned payload should have a `referenceID` for any payment service items.
1. Save the following payload to `get_payment_request_edi.json`, replacing `<paymentRequestID>` with the payment request ID that was output from the previous step:
    ```
    {
      "paymentRequestID": "<paymentRequestID>"
    }
    ```
1. Get the EDI for that payment request:
    ```
    go run ./cmd/prime-api-client --insecure support-get-payment-request-edi --filename get_payment_request_edi.json | jq -r '.edi'
    ```
    You can use this same `prime-api-client` (with appropriate flags) to view the EDI in our deployed environments.
1. Verify that you see the reference ID in the returned EDI in the `N9` segments.  There should be four of them, one for each service item.

If desired, you can look at other `prime-api-client` commands like `fetch-mto-updates` that return payment service items if they exist and verify `referenceID` appears for them as well.  There are also GHC endpoints that return payment service items that you could view through the browser inspector in the TIO interface.

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#logging)
* [ ] The requirements listed in [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely) have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #g-database
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-3718) for this change

### CODE REVIEW

**Review by gidjin — APPROVED** [review]
Changes look good. I left a question for better understanding but verified it's working locally as expected.

**Review by pearl-truss — APPROVED** [review]
I was able to follow the steps in the PR description and get the expected EDI with PSI reference ids in the N9 fields

File: pkg/models/payment_service_item.go [hunk 1]
```diff
@@ -72,5 +80,79 @@ func (p *PaymentServiceItem) Validate(tx *pop.Connection) (*validate.Errors, err
...
+	// Get a version of the PSI's ID without dashes.
+	psiID := fmt.Sprintf("%x", p.ID)
+	psiIDLength := len(psiID)
+
+	// We want at least 8 hex digits of the ID, but no more than the max allowed by the EDI (if needed for uniqueness).
+	currentLength := PaymentServiceItemMinReferenceIDSuffixLength
+	maxSuffixLength := PaymentServiceItemMaxReferenceIDLength - len(*mto.ReferenceID) - 1 // including the hyphen
+	if maxSuffixLength > psiIDLength {
+		maxSuffixLength = psiIDLength
+	}
```
[the remark this answers is not in the document]
Comment (reggieriser) [reply pkg/models/payment_service_item.go:136]: Right.  Once I introduced the `PaymentServiceItemMaxReferenceIDLength` constant, I was trying to ensure the constant value could change and the rest of the code would still just work.  This is mainly just a safety to make sure that I don't index past the end of the psiID string with `psiID[:currentLength]` and cause a panic.

**Review by Ryan-Koch — APPROVED** [review]
DB things look good

**Pull request merged**